connectors/kafka/src/main/java/forklift/controller/KafkaController.java

```
@@ -0,0 +1,349 @@
+    /**
+     * Acknowledge that processing is beginning on a record.  True is returned if this controller is still managing the
+     * partition the record originated from.  If the partition is no longer owned by this controller, false is returned
+     * and the message should not be processed.
+     *
+     * @param record the record to acknowledge
+     * @return true if the record should be processed, else false
+     * @throws InterruptedException if the thread is interrupted
+     */
+    public boolean acknowledge(ConsumerRecord<?, ?> record) throws InterruptedException {
+        AtomicInteger flowCount = flowControl.get(new TopicPartition(record.topic(), record.partition()));
+        if (flowCount != null) {
+            int counter = flowCount.decrementAndGet();
+            if (counter == 0) {
+                synchronized (this) {
+                    this.notifyAll();
+                }
+            }
+        }
+        log.debug("Acknowledge message with topic {} partition {} offset {}", record.topic(), record.partition(), record.offset());
+        return running && this.acknowlegmentHandler.acknowledgeRecord(record);
+    }
+
+    /**
+     * Starts the controller.  Must be called before any other methods.
+     * <p>
+     * <strong>WARNING: </strong>Avoid starting a service which has already run and been stopped.  No attempt is made
+     * to recover a stopped controller.
+     */
+    public void start() {
+        running = true;
+        executor.submit(this::controlLoop);
+    }
+
+    /**
+     * Stops the controller.
+     *
+     * @param timeout  the time to wait for the service to stop
+     * @param timeUnit the TimeUnit of timeout
+     * @throws InterruptedException if the thread is interrupted
+     */
+    public void stop(long timeout, TimeUnit timeUnit) throws InterruptedException {
+        executor.shutdownNow();
+        kafkaConsumer.wakeup();
+        running = false;
+        executor.awaitTermination(timeout, timeUnit);
```

This has a result value indicating whether awaiting termination succeeded - maybe we want to log an error or throw an exception when the unexpected happens.

I would expect that this almost always succeeds, so wouldn't we want to know when it doesn't?